### PR TITLE
Remove data folding from model architecture

### PR DESCRIPTION
**Pull request by ermanok**
- Folding added as a dataset transformation
- Unfolding function added and called in train.py when required
- UNET models updated
- CamVid and AISegmnet datasets updated to call new folding transformation

### CODE REVIEW

File: ai8x.py [hunk 1]
```diff
@@ -33,6 +33,52 @@ def __call__(self, img):
...
+
+    for i in range(fold_ratio):
+        for j in range(fold_ratio):
+            ch_index_start = num_out_channels*(i*fold_ratio + j)
+            ch_index_end = num_out_channels * (i*fold_ratio + j + 1)
+            img_batch_uf[:, :, i::fold_ratio, j::fold_ratio] = \
+                img_batch[:, ch_index_start:ch_index_end, :, :]
+
+    return img_batch_uf
+
```
Comment (seldauyanik-maxim) [inline ai8x.py:88]: May the unfolding function for a single image be helpful to expose and utilize it also in here?
For example usages in notebooks etc.

File: datasets/camvid.py [hunk 2]
```diff
@@ -90,21 +90,21 @@ def __init__(self, root_dir, d_type, classes=None, download=True, transform=None
...
-                                img_crop_subsample = img_crop[i::fold_ratio, j::fold_ratio, :]
-                                if img_crop_folded is not None:
-                                    img_crop_folded = np.concatenate((img_crop_folded,
-                                                                      img_crop_subsample), axis=2)
-                                else:
-                                    img_crop_folded = img_crop_subsample
-                        self.img_list.append(img_crop_folded)
-                        self.lbl_list.append(lbl_crop)
+                    # if fold_ratio == 1:
+                    self.img_list.append(img_crop)
```
Comment (seldauyanik-maxim) [inline datasets/camvid.py:94]: May it better be to remove these parts instead of commenting them out

**Review by seldauyanik-maxim — DISMISSED** [review]

File: ai8x.py (hunk 1 above)
Comment (ermanok) [reply ai8x.py:88]: The unfolding step is moved just after having the model output. Therefore, we got the output in batches even when the batch_size=1. But I can add an unfold function for a single image as well, if you think it is necessary.

File: datasets/camvid.py (hunk 2 above)
Comment (ermanok) [reply datasets/camvid.py:94]: It has already done in one of the latest commits. Did you write this comment on a previous commit?

File: ai8x.py [hunk 3]
```diff
@@ -33,6 +33,60 @@ def __call__(self, img):
...
+    Fold data to increase the number of channels. An interlaced approach used in this folding
+    as explained in [1].
+
+    [1] https://arxiv.org/pdf/2203.16528.pdf
+    """
+    def __init__(self, fold_ratio):
+        self.fold_ratio = fold_ratio
+
+    def __call__(self, img):
+        img_folded = None
```
Comment (gorkemulkar) [inline ai8x.py:47]: I think it'd be good to do nothing and return faster when fold_ratio==1. As it was implemented for the unfold.

File: models/ai85net-unet.py [hunk 4]
```diff
@@ -228,21 +228,21 @@ def forward(self, x):  # pylint: disable=arguments-differ
...
+        #                           dec0.shape[3]*self.fold_ratio), dtype=dec0.dtype,
+        #                          device=t_device, requires_grad=False)
+        #     for i in range(self.fold_ratio):
+        #         for j in range(self.fold_ratio):
+        #             ch_index_start = self.num_classes*(i*self.fold_ratio + j)
+        #             ch_index_end = self.num_classes*(i*self.fold_ratio + j + 1)
+        #             dec0_u[:, :, i::self.fold_ratio, j::self.fold_ratio] = \
+        #                 dec0[:, ch_index_start:ch_index_end, :, :]
+
+        #     return dec0_u
```
Comment (gorkemulkar) [inline models/ai85net-unet.py:245]: Here I see some commented out code.

File: ai8x.py [hunk 5]
```diff
@@ -33,6 +33,60 @@ def __call__(self, img):
         return img.sub(0.5).mul(256.).round().clamp(min=-128, max=127).div(128.)
 
 
+class fold:
```
Comment (gorkemulkar) [inline ai8x.py:36]: I think we need to CamelCase the class names according to PEP8 format. "normalize" needs to be changed too. 

Surely this is not critical, we can change them later in another commit.

**Review by gorkemulkar — CHANGES_REQUESTED** [review]

File: ai8x.py (hunk 5 above)
Comment (ermanok) [reply ai8x.py:36]: I think it is better to fix all in another commit. It requires a big regression test.

File: ai8x.py (hunk 1 above)
Comment (seldauyanik-maxim) [reply ai8x.py:91]: Sure, understood. There may be one exceptional use case where data loaders can be loaded directly in some of the notebooks wo torch data loader creation losing batch dimension. Not a critical case though.

File: datasets/camvid.py (hunk 2 above)
Comment (seldauyanik-maxim) [reply datasets/camvid.py:94]: Yes many thanks, the last commit addresses all.

**Review by MaximGorkem — APPROVED** [review]
Looks better, thanks

**Review by alicangok — APPROVED** [review]

File: ai8x.py [hunk 6]
```diff
@@ -33,6 +33,63 @@ def __call__(self, img):
...
+
+    t_device = img_batch.get_device()
+    if t_device < 0:
+        t_device = None
+
+    num_out_channels = img_batch.shape[1] // (fold_ratio*fold_ratio)
+
+    img_batch_uf = torch.zeros((img_batch.shape[0], num_out_channels,
+                                img_batch.shape[2]*fold_ratio, img_batch.shape[3]*fold_ratio),
+                               dtype=img_batch.dtype, device=t_device, requires_grad=False)
```
Comment (alicangok) [inline ai8x.py:81]: Would getting rid of lines 73-75, and changing line 79 to

img_batch_uf = torch.zeros( ... , device=img_batch.device, ... )

work?

Comment (ermanok) [reply ai8x.py]: This change has been applied. Thanks.

**Review by alicangok — APPROVED** [review]
Looks good.
